### DISCUSSION TITLE
psycopg 3 connection pool versus sqlalchemy connection pool

### DISCUSSION THREAD

**Discussion by billcrook** [sqlalchemy/sqlalchemy#7478] [Usage Questions]
With more advanced features coming in the psycopg 3 connection pool, I'm wondering:
* What are the benefits of using one pool over the other?
* Is there any interaction between the 2 pool implementations?
* Is is possible to use sqlalchemy with psycopg 3 connection pool?
* Is it still recommended to use sqlalchemy connection pool?

Interested to hear comments from @dvarrazzo as well. Thanks

**Comment by zzzeek** [comment]
> With more advanced features coming in the psycopg 3 connection pool, I'm wondering:
> 
>     * What are the benefits of using one pool over the other?

im not familiar with psycopg3's pool, but if pooling performance is super important to you, and your application runs with more than one Python process (very likely if performance matters), you'd be better off using pgbouncer.  that is, the main "pool" issue is that Python processes that all have their own pool don't share the connections as efficiently as if you have high performance middleware handling it.

> 
>     * Is there any interaction between the 2 pool implementations?

looking at https://www.psycopg.org/psycopg3/docs/advanced/pool.html it looks like psycopg3's pool is thankfully not implicit, you use a pool object explicitly, so out of the box, psycopg3's pool would not be used, so no.

> 
>     * Is is possible to use sqlalchemy with psycopg 3 connection pool?

sure, you would set up for NullPool in your create_engine and then write a creator function that uses the pool.   Here is an example that is doing this identical thing for cx_Oracle's connection pool, you can use this as is, just replace the cx_Oracle functions for the psycopg3 ones:  https://docs.sqlalchemy.org/en/14/dialects/oracle.html#using-cx-oracle-sessionpool

that example is there because folks wanted to use that pool, I think cx_Oracle's pool might be tapping into some more enterprise-level Oracle server stuff (like middleware) that would make it worth it.

> 
>     * Is it still recommended to use sqlalchemy connection pool?

SQLAlchemy is designed around the assumptions used by this pool so for most predictable results, mostly in the area of what happens when connections are killed by the server and similar areas, you'd use the built-in pool.  

Looking at psycopg3's docs, looks like some pretty useful things are lacking:

> The health of the connection is not checked when the pool gives it to a client.

ouch. that's a basic one.  pgbouncer has this, SQLAlchemy has this for every DB ([pool_pre_ping](https://docs.sqlalchemy.org/en/14/core/pooling.html#disconnect-handling-pessimistic)), im pretty sure Hibernate's pools have this, if psycopg3 wants to provide a connection pool it should be adding this.

> The health of the connection is not checked when the connection is in the pool.

SQLAlchemy's pool does not do this either but we have a [pool_recycle](https://docs.sqlalchemy.org/en/14/core/pooling.html#setting-pool-recycle) option that helps a lot with this, though the pre-ping option prior is mostly all you need.


> 
> 
> Interested to hear comments from @dvarrazzo as well. Thanks

**Comment by edwardzjl** [comment]
I'm experiencing a similar issue.
In my project, I have a third-party dependency that relies on psycopg, so I created a separate psycopg connection pool for it.
Initially, I ended up with two connection pools: one for psycopg and another for SQLAlchemy. This setup is cumbersome, so I attempted to pass psycopg connections directly into SQLAlchemy using the following approach:

```python
sqlalchemy_engine = create_async_engine(
    url=settings.db_url,
    async_creator=pool.getconn,
    poolclass=NullPool,
)
```

However, this code doesn’t work as expected. According to the [API doc](https://www.psycopg.org/psycopg3/docs/api/pool.html#psycopg_pool.ConnectionPool.getconn), I need to call `putconn()` after using the connection. As a result, after making `pool_size` requests, my application gets stuck because no more connections are available.




Additionally, regarding the statement:

> The health of the connection is not checked when the pool gives it to a client.

This appears to be outdated. psycopg now includes a `check` parameter to validate the health of a connection before handing it off to the client.

**Reply by dvarrazzo** [reply]
> It seems strange that SQLAlchemy offers `async_creator` but not `async_destructor` or something like that, but yes, skimming their docs I don't see that. Maybe you can hear from that project and ask them what is the intended usage.

I assume that they assume that they can just call `async close()`. If so maybe you can make your own connection subclass that on `close()` would call `await self._pool.putconn(self)` instead? `_pool` is not part of the public interface but let's not tell anyone that you are using it.

>> The health of the connection is not checked when the pool gives it to a client.

> This appears to be outdated. psycopg now includes a check parameter to validate the health of a connection before handing it off to the client.

Correct, thank you for pointing it out. Since psycopg 3.2 it's possible to configure some connection check.

**Reply by CaselIT** [reply]
@edwardzjl you can use an event to put a connection back to the psycopg pool.

see:
- https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.PoolEvents.checkin
- https://docs.sqlalchemy.org/en/20/core/pooling.html#sqlalchemy.pool.ConnectionPoolEntry.driver_connection
- https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#using-events-with-the-asyncio-extension

**Reply by edwardzjl** [reply]
Thank you!

**Reply by chrispy-snps** [reply]
I am trying to adapt psycopg 3's `ConnectionPool` in SQLAlchemy 2.0. (See [this psycopg post](https://github.com/psycopg/psycopg/discussions/994) for details - @dvarrazzo, thanks for your help there!)

Unfortunately the second transaction hangs instead of completing. Here's what I've got:

```python
import time
import logging

logging.basicConfig(level=logging.DEBUG, format="%(asctime)s %(levelname)s %(name)s: %(message)s")
logging.getLogger("sqlalchemy.pool").setLevel(logging.DEBUG)

import sqlalchemy
from psycopg_pool import ConnectionPool

db_url = "postgresql+psycopg://postgres:postgres@hostname:port/database"

# create a psycopg_pool connection pool
pool = ConnectionPool(
    conninfo=db_url.replace("+psycopg", ""),
    min_size=0,
    max_size=1,
    max_idle=5,
)

# create a SQLAlchemy engine using the psycopg_pool connection pool
engine = sqlalchemy.create_engine(
    url=db_url,
    poolclass=sqlalchemy.pool.NullPool,  # disable SQLAlchemy's default connection pool
    creator=pool.getconn,                # use the psycopg_pool connection pool instead
)

# generate some "activity"
def do_stuff():
    with engine.connect() as connection:
        result = connection.execute(sqlalchemy.text("SELECT now()"))
        print(f"Current time: {result.fetchone()[0]}")

do_stuff()
do_stuff()

# now go idle
# (and wait for max_idle timeout to release connection)
time.sleep(15)
```

I expected the `with` context manager to release the first pool connection automatically. Do I need to pass a different function as the `creator` for this to work?

Thanks!

**Reply by CaselIT** [reply]
please see the event mentioned here https://github.com/sqlalchemy/sqlalchemy/discussions/7478#discussioncomment-10907783

**Reply by chrispy-snps** [reply]
@CaselIT - thanks! I did look at those posts, and it seems they provide the solution given @edwardzjl's positive reaction. However, I was unable to understand how to craft a solution from them.

If that means I'm not ready to be using these mechanisms in my own code, I accept that too. :)  Perhaps I'll disable connection pooling for now, then revisit this topic when my knowledge of SQLAlchemy has increased.

**Reply by chrispy-snps** [reply]
@edwardzjl - did you get psycopg 3's `ConnectionPool` working in SQLAlchemy?

**Reply by edwardzjl** [reply]
No I switched to pgbouncer instead.
